src/tools/isabelle_executor.py

```diff
@@ -0,0 +1,586 @@
+    proof_context_match = re.compile(proof_context_regex, re.MULTILINE)
+
+    # Matches theory initialization
+    begin_theory_regex = r"theory([\s\S]*)imports([\s\S]*)begin"
+    begin_theory_match = re.compile(begin_theory_regex, re.MULTILINE)
+
+    def __init__(self, project_root: str = None, main_file: str = None, use_hammer: bool = False, timeout_in_sec: int = 60, 
+                 use_human_readable_proof_context: bool = False, proof_step_iter: typing.Iterator[str] = None, 
+                 suppress_error_log: bool = False):
+        assert proof_step_iter is None or isinstance(proof_step_iter, typing.Iterator), \
+            "proof_step_iter must be an iterator"
+        assert main_file is not None or proof_step_iter is not None, \
+            "Either main_file or proof_step_iter must be provided"
+        assert (main_file is None) or (project_root is not None), \
+            "Project root must be provided for Isabelle, if main file is provided"
+        assert main_file is None or (os.path.exists(main_file) and main_file.endswith(".thy")), \
+            "main_file must be a valid path to a '.thy' file"
+        assert project_root is None or (os.path.exists(project_root) and os.path.isdir(project_root)), \
+            "project_root must be a valid path to a directory"
+        self.use_human_readable_proof_context = use_human_readable_proof_context
+        self.project_root = project_root if project_root is not None else "."
+        self.main_file = main_file
+        self.use_hammer = use_hammer
+        self.timeout_in_sec = min(timeout_in_sec, 120) # Maximum 120s timeout
+        self.current_stmt = None
+        self.line_num = 0
+        self.current_state = 0
+        self.line_num_to_state = {}
+        self.main_file_iter = proof_step_iter
+        self.buffer = ""
+        self.suppress_error_log = suppress_error_log
+        self.pisa_env : PisaEnv = None
+        self.proof_context : ProofContext = None
+        self.curr_lemma_name : typing.Optional[str] = None
+        self.curr_lemma : typing.Optional[str] = ""
+        self._proof_running = False
+        self._top_level = True
+        self.local_theorem_lemma_description: typing.OrderedDict[str, str] = OrderedDict()
+        self.execution_complete = False
+        self.global_lemmas = []
+    
+    def __enter__(self):
+        self._all_dep_handles = []
+
+        if self.main_file_iter is None:
+            self.main_file_iter = IsabelleLineByLineReader(self.main_file).instruction_step_generator()
+
+        # PISA clients must provide a file and working directory. If these are not provided, 
+        # use the default header, which may or may not be sufficient.
+        if self.main_file is None:
+            logger.warning("Initialising Isabelle environment with default theory header and imports (Complex_Main). Pass in a file and project root to import additional theories")
+            self.pisa_env = initialise_env()
```

Comment:
I think PISA requires that all theories be placed in a file. If the user doesn't provide one, I use a default, but that may not have all the imports necessary. Since all the benchmarks are in their own file anyway I don't really anticipate any problems